### DISCUSSION TITLE
Support Custom Public Properties

### DISCUSSION THREAD

**Discussion by jonassiewertsen** [livewire/livewire#3812] [3. Feature ideas]
What do you think about the idea to support custom public properties? **I would like to create a PR for it**, but would like some feedback first.

With version 2.6 there came a great new feature: Wirable Properties:
https://laravel-livewire.com/docs/2.x/properties#wireable-properties

Being able to implement an Interface does mean, that we need to be in control of all the Data we are using, but this is not always the case, if fx working with other Packages. 

_My Example class names have not been thought through yet, but should be fine for a first rough sketch:_

```php
// Something like a Custom Property Manager which can be bound via a Service Provider.

// Entry class is a imaginary Class from a Laravel Package (like for example in statamic.com)
// EntryClassResvoler is your custom class, which should simply implement the Wireable Interface to make sure the class type can be hydrated and dehydrated.
CustomPropertyManager::register(Entry::class, EntryClassResolver::class)
```

Implementing it, would mean to extend the HydratePublicProperties middelware, so we can check if the `CustomPropertyManager` can hydrate or dehydrate a specific class. It would need something like this:

```php
else if (CustomPropertyManager::hasType($value)) {
```

Here
https://github.com/livewire/livewire/blob/master/src/HydrationMiddleware/HydratePublicProperties.php#L40-L59
And here:
https://github.com/livewire/livewire/blob/master/src/HydrationMiddleware/HydratePublicProperties.php#L91-L118

_This would be amazing, as you could use the Wireable Interface as implemented right now OR could as well register Custom Properties in your application. 
In some cases the Wireable interface can not be applied, which would be solved by this idea. Depending on your Application, if having lot's of custom classes, my idea might help keeping the dehydration and hydration in a separate place (separation of concerns), which might be a big help to some applications. In the end this would make Livewire even more flexible for custom properties_ 

**Is this a good idea and worth creating a PR for it?**

**Comment by mabdullahsari** [comment]
I mean, can't you just use custom casters for that instead of re-inventing the wheel?

**Reply by edalzell** [reply]
What if they're not models (like in this case?). The use case here is to find a way to "attach" the wireable to an EXISTING class/object, one we don't have control over.

**Comment by joshhanley** [comment]
@jonassiewertsen I think it could be good!

We could actually clean up all the existing options to use that too (like `wireable`, `carbon`, etc.), which would make that area neater rather than just a bunch of conditionals.

One thing to consider, is if there is a "property manager registry", we would need to ensure each class only can have one resolver assigned to it, so there isn't an issue of multiple resolvers and which one comes first.

Will be good to get Caleb's thoughts on it before you implement it, I'll see if he can have a look.

**Reply by jonassiewertsen** [reply]
I thought about cleaning up as well, but didn't want to talk about to many things in one idea. 
The following way might be a good solution:

Step 1: 
Cleaning up. Instead of Conditions use some kind of manager, by refactoring into own classes for each case (as you did write) 

Step 2:
Implement the idea of this discussion, after step 1 is approved and implemented.

Splitting it up like that will make it much more easy to approve those PRs. 

I would love to contribute. Let me know and I'm on it 🚀

Am excited to hear Caleb's thoughts on this.

**Reply by calebporzio** [reply]
Yeah, I like the concept of centralizing all this logic and cleanup up the massive if/else statement in HydratePublicProperties

This is a core part of the Livewire system however, so great care will need to be given to a refactor.

I'm in favor, but hesitant to touch such a key part of the system - so maybe if someone attempts this, just be careful and document things well.

**Reply by joshhanley** [reply]
@jonassiewertsen I've just recently ripped a lot of that class apart and rebuilt it to add deep nesting support in, so happy to help if you need it 🙂

Maybe we can do it in the reverse order of what you suggested? Create a manager and implement it as another conditional for now, then we can progressively move the other types over to the new manager style structure once it has been implemented and throughly tested with non-core types.

That way we can approach each change in a modular fashion and throughly test each change/ migration.

**Reply by jonassiewertsen** [reply]
Hej Josh

That does sound good. I made the first sketch for a refactor, which I need to adjust a lot and would like some feedback on. I'll reach out on Discord.

**Comment by joshhanley** [comment]
@jonassiewertsen so I've been thinking about this a lot, and I believe there is one extra issue that we probably should handle as part of this.

Currently, there is hydrate/dehydrate of properties which we are hoping to be able to register custom handlers for, as discussed here.

But one area it doesn't cover is updating of properties.

A good example of this, is a carbon instance.

If we have in a component, a public property that is type hinted to Carbon, it gets hydrated/dehydrated happily, but if you try to update that property (with a datestring) from the frontend, it breaks.

```php
public ?Carbon $date;
```

![image](https://user-images.githubusercontent.com/882837/135552088-1209643c-3e1a-4042-9e28-51299816e14a.png)

Currently to work around this, I'm using an updating hook with the value being passed in by reference, so I can tell Carbon to parse the value myself.

```php
public function updatingDate(&$value) {
    if ($value == '') return $value = null; // Need this empty string check to set to null if the text input is empty.
    
    $value = Carbon::parse($value);
}
```

So I'm wondering if the handler should also have a method for handling updates from the front end. That would allow us to define exactly how a properties value should be parsed to become that specific object.

I'd imagine then we would have an interface for the handler that has 

```php
interface PropertyHandler {
    public function hydrate($value);

    public function dehydrate();

    public function update($value);
}
```

This means we would only need to define how to handle an update to a Carbon property once, on the handler, that way any Carbon properties will use that to process updates.

**Reply by edalzell** [reply]
love this

**Reply by jonassiewertsen** [reply]
That does make a lot of sense to me! Let's do it that way.

I already started working on a PR, but have been a little short on time.

**Reply by joshhanley** [reply]
@jonassiewertsen Ok cool 🙂 let me know if you need me to do anything/ want to look over anything

**Comment by iksaku** [comment]
Jumping in here... Just posted [a similar discussion](https://github.com/livewire/livewire/discussions/4080) for which this would provide a super integration!
Would love to help you @jonassiewertsen if you have any kind of proof of concept :)

**Reply by jonassiewertsen** [reply]
Cool! I need to finisch my draft that the first smoke test is done. After that, I would love some help.

**Comment by joshhanley** [comment]
Here is @jonassiewertsen current diff of the feature https://github.com/livewire/livewire/compare/master...jonassiewertsen:add-public-property-manager?expand=1

And I sent him some comments on it, which I will add below (copied verbatim):

---

Looking good so far!

I think rename it to LivewirePropertyManager, as then when you want to register custom bindings, say in app service provider, it will just be LivewirePropertyManager::register(Carbon::class, CarbonResolver::class); or something like that.

My first thought was bind instead of register but I think we should match Laravel's service container naming (whatever that is), so it's consistent.

Also the bind/register should allow overwriting (I think you have it set so it already does), but we should add a test to ensure that.

I'd move the hydration middleware back to where it was, and move the property manage class to the src directory. So that way the import is use Livewire\LivewirePropertyManager; (makes the changes in the PR smaller too 🙂)

If ->get($class) can't be found, I think return null that way we don't have to do error handling as we are changing properties over, instead if it's null we can just fall back to the next option (an existing implemenation in the big conditional).

On that, I think the property manager should be at the top of the conditionals in HydratePublicProperties, so it gets checked first, then if nothing exists for a class, it starts to run down the existing list of supported types. This will allow us to progressively change properties over to their own resolver classes. Plus users will be able to register a resolver straight away to override the defaults.

To implement the changes in hydrate/dehydrate, we will probably need to use reflection to get the properties type, which then get's passed to the ->contains method. Maybe that should be called ->has()?

 Then in the hydrate method for example it would be data_set($instance, $property, LivewirePropertyManager::get($type)->hydrate($value)).
This part I'm not a 100% sure on. Currently Livewire has dataMeta for collections/models/ etc. I'm wondering whether this should still be used or whether the metadata should be co-located with the actual data, as that will change how the dehydrate method is used, as it may need the metadata collections passed into it or something.

Also I'm wondering whether the wireable interface should be used, or whether there should be a new one for PropertyResolver or something that has hydrate,dehydrate,update in it.

I also thinking that interface should also have a validate/validation method as well. The reason for that is properties need to be serialised to an array (I think) for validation to work properly on them, see this "Validation DTO's" PR that adds support for validating wireable DTOs https://github.com/livewire/livewire/pull/4036 

Actually the reflection probably won't work, we will need to store that in a properties metadata like you have on dehydration, that way it supports it if people don't type hint properties. But that then brings up the question of should think support non-type hinted properties? If it does support non-type hints, then on update, is the property locked to a specific type due to what it was on the initial dehydration. What if it was null, then the system won't know the update coming in is a carbon instance for example, and so won't know to use the carbon resolver. Maybe we should only support properties with type hints for this to work

**Reply by jonassiewertsen** [reply]
Thanks a lot for the great feedback @joshhanley 

I do have more time this week, so I will continue my work on this PR.

**Reply by joshhanley** [reply]
@jonassiewertsen no worries! 🙂

Let me know if you need anything!

**Comment by joshhanley** [comment]
Have been discussing this feature with @stancl as he was originally looking for a way to use the `wireable` interface in his package, except that Livewire isn't a requirement, so didn't want his object to break if the `wireable` interface doesn't exist.

We discussed the idea of using a custom resolver for his object instead, so it would implement the resolver interface instead.

Which then in his package service provider, he could do a class exists check for Livewire, and only if it exists, then bind his custom resolver in. That way the package will still function happily if Livewire isn't installed and his original object class isn't impacted (or even have any custom Livewire implementation in it for that matter).

From that though we went through the idea of using a soft interface (using method exists) in the property manager implementation though, so that way resolvers could implement some of the methods and not all. Like if an object is already `arrayable`, then it wouldn't need a `dehydrate` method, just a `hydrate` method for example.

We would still provide an interface that has all the methods if people want it so they can implement the methods using their IDE.



One other question that came up though is, if we have an Eloquent Model resolver, but then also had a custom resolver for a specific model type like `PostModelResolver`, would there be a way to check for and use the `PostModelResolver` over the `EloquentModelResolver`?

Not sure how to handle that, but it would be interesting!

**Reply by jonassiewertsen** [reply]
After a little testing, I will create that PR so we can discuss it there. Some thoughts I do have on this: 

**Soft interface etc**

If registering a custom property, I did change the interface to an abstract class:
https://github.com/livewire/livewire/compare/master...jonassiewertsen:add-public-property-manager?expand=1#diff-e9821fa6bb4d4d7394a38a4245dbf048c1f84240cd817e5bd30135fae4f16c19R5

This might be the right way, to solve problems as described by @joshhanley 

I did it, as we sometimes might need (as with a Carbon instance) an update method, but I did not want everybody implementing it, as many might not need it at all.
That would make it easier to check for certain things. Anyways, I am not sure if it is useful making the Property Handler too smart though?

> One other question that came up though is, if we have an Eloquent Model resolver, but then also had a custom resolver for a specific model type like PostModelResolver, would there be a way to check for and u**se the PostModelResolver over the EloquentModelResolver**?

Not properly tested yet, but it should already work that way. The custom property resolver is running nearly as the first check:
https://github.com/jonassiewertsen/livewire/blob/f119d15648c0df154b075449913700056a8b4bb4/src/HydrationMiddleware/HydratePublicProperties.php#L100

By doing so: If you did register your own custom class, it will match and be matched as a custom property instead of a normal collection. If no custom Model has been registered, everything will work as before.

**Reply by joshhanley** [reply]
Looking good! 

Can you also add validation in as well - see this PR for Wireable #4036 - as the validation system needs the objects in certain format (array I think) to work properly, so we will need a way to allow custom properties to implement a `validate` method (even if that is just deferring to the `dehydrate` method).

https://github.com/livewire/livewire/blob/4a349459fd9c8e2bbeae2306d7fd97ba424de006/src/ComponentConcerns/ValidatesInput.php#L305-L310

---

I'm personally not a fan of the abstract class and don't think it should be enforced, instead just doing a method exists check on the resolver would be better. That way, if update is not implemented for example, then there is no issues and just Livewire's default update process will be followed.

My thoughts were that if needed then you could just implement a hydrate method, or just an update method, whatever is required for that specific property. Any of the other methods would just fall through and keep working down the conditionals in `HydratePublicProperties.php`.

I think we can still have an interface though that has all 4 methods (hydrate, validate, update, dehydrate) if people want to use it, but that would mostly be for IDE benefit, so it you can just click implement and it will pre-populate the method stubs for you.

---

Re: the eloquent model stuff, maybe it would be good, when doing the check for existance and returning the resolver, that we check the array in reverse order. That way later bindings have preference over earlier ones. That way if you have a custom resolver for eloquent models bound first, and then a specific one for `PostResolver` then the post resolver is checked first.


Great job! Loving this feature and your implementation looks so clean. Looking forward to being able to use it! 😁

**Reply by jonassiewertsen** [reply]
Ohhh ... I just created a first draft PR, so we can comment it there and did not see your comment. Should we move our discussion to the PR itself? 

https://github.com/livewire/livewire/pull/4154

Thanks for those nice words and the great feedback!

**Reply by joshhanley** [reply]
@jonassiewertsen yep I think I hit send as you posted your PR haha!

Will copy my comment to the PR 🙂
